sql/log.cc

```diff
@@ -163,6 +163,111 @@ static SHOW_VAR binlog_status_vars_detail[]=
   {NullS, NullS, SHOW_LONG}
 };
 
+/**
+  This class implementes the feature to rename a binlog cache temporary file to
+  a binlog file. It is used to avoid holding LOCK_log long time when writting a
+  huge binlog cache to binlog file.
+
+  With this feature, temporary files of binlog caches will be created in
+  BINLOG_CACHE_DIR which is created in the same directory to binlog files
+  at server startup.
+*/
+class Binlog_free_flush
+{
+public:
+  Binlog_free_flush() {}
+
+  /**
+    Check whether free flush should be executed on the cache_data.
+
+    @param group_commit_entry object of current transaction
+
+    @retval true    it should do free flush
+    @retval false   it should do normal commit.
+  */
+  bool should_free_flush(const MYSQL_BIN_LOG::group_commit_entry *entry) const;
+
+  /**
+    This function is the entry function to do free flush. It first,
+    rotate the binlog, then rename the temporary file of the
+    binlog cache to new binlog file, after that it commits the transaction.
+
+    @param entry, group_commit_entry object of current transaction.
+
+    @retval true    free flush succeeds.
+    @retval false   free flush fails, it should go to normal commit process.
+  */
+  bool commit(MYSQL_BIN_LOG::group_commit_entry *entry);
+
+  /**
+    After rotate has created the new binlog file, it copies the content
+    of the new binlog file to the binlog cache, delete the new binlog file
+    and then rename the binlog cache to the new binlog file.
```

Comment:
Perhaps here, we can be specific about "the content" as to the actual binlog events that are copied.

Reply:
I didn't write the specific events down, because the function doesn't care about the exact events. It just do bytes copy. Even new events added in future, this function doesn't need to be changed.

Reply:
Right.. Though even without being specific, I think the comment should provide an expectation for the content of the rotated binlog file (i.e. at what point in rotation). Perhaps something like
> During binlog rotate, after creating the new binlog file and writing the events that describe its state (e.g. Format description event)}, copy ...

Reply:
fine